### PR TITLE
Clarify Quantification usage and functions

### PR DESCRIPTION
**Pull request by abailly-iohk**
Provide some Haddock comments on variables quantification functions and also require to provide predicate to `withGenQ` smart constructor.

Checklist:
- [x] Check source-code formatting is consistent

### CODE REVIEW

File: quickcheck-dynamic/src/Test/QuickCheck/DynamicLogic/Quantify.hs [hunk 1]
```diff
@@ -32,12 +32,21 @@ import Test.QuickCheck
...
+-- * A _shrinking_ strategy for generated values in the case of a
+--   failures ensuring they stay within the domain,
+-- * A _predicate_ allowing finer grained control on generation
+--   and shrinking process, e.g in the case the range of the generator
+--   depends on trace context.
 data Quantification a = Quantification
-  { genQ :: Maybe (Gen a)
+  { -- FIXME: why is this a Maybe? It makes things annoying down the road, smells like
+    -- something which would benefit from better typing?
+    genQ :: Maybe (Gen a)
```
Comment (UlfNorell) [inline quickcheck-dynamic/src/Test/QuickCheck/DynamicLogic/Quantify.hs:49]: This is to allow empty generators. A problem that you often run into with QuickCheck is `elements` crashing on an empty list. Having a `Maybe` here allows `elements` to succeed (but with an empty generator).

This lets you write nice code like this (from the registry example):
```
pickAlive :: DL RegState (Var ThreadId)
pickAlive = do
  alive <- getAlive
  forAllQ $ elementsQ alive
```
There is no need to check if there are any alive threads here. If there aren't, we'll get an empty generator and the DL machinery will prune that branch.

**Review by UlfNorell — CHANGES_REQUESTED** [review]

**Comment by MaximilianAlgehed** [comment]
Could we rename `withGenQ` to something that indicates that it's not quite safe?

**Comment by abailly-iohk** [comment]
@UlfNorell @MaximilianAlgehed Are you ok with this change? If yes, we could merge and then cut a 4.0.0 version with the negative testing stuff and this small improvement

File: quickcheck-dynamic/src/Test/QuickCheck/DynamicLogic/Quantify.hs [hunk 2]
```diff
@@ -51,10 +63,9 @@ generateQ q = fromJust (genQ q) `suchThat` isaQ q
 shrinkQ :: Quantification a -> a -> [a]
 shrinkQ q a = filter (isaQ q) (shrQ q a)
 
--- | Wrap a `Gen a` generator in a `Quantification a`.
--- Uses given shrinker.
-withGenQ :: Gen a -> (a -> [a]) -> Quantification a
-withGenQ gen = Quantification (Just gen) (const True)
+-- | Construct a `Quantification a` from its constituents.
```
Comment (MaximilianAlgehed) [inline quickcheck-dynamic/src/Test/QuickCheck/DynamicLogic/Quantify.hs:66]: I would like to see some better documentation here explaining that it is important for the `(a -> Bool)` to be sound and complete w.r.t. the generator.

**Review by MaximilianAlgehed — CHANGES_REQUESTED** [review]

File: quickcheck-dynamic/src/Test/QuickCheck/DynamicLogic/Quantify.hs (hunk 2 above)
Comment (abailly-iohk) [reply quickcheck-dynamic/src/Test/QuickCheck/DynamicLogic/Quantify.hs:66]: I think this is documented in the `Quantification` haddocks

Comment (MaximilianAlgehed) [reply quickcheck-dynamic/src/Test/QuickCheck/DynamicLogic/Quantify.hs:66]: But that means the user has to read that documentation to use this function correctly - isn't that a shame?

Comment (MaximilianAlgehed) [reply quickcheck-dynamic/src/Test/QuickCheck/DynamicLogic/Quantify.hs:66]: When I think about it, one way to make sure it only has to be sound (in the sense of only being True on things that could have been generated) is to use `suchThat` on the Gen. That way the user doesn't have to ensure that it is True on _all_ things that were generated, that will be the case automatically.

**Comment by abailly-iohk** [comment]
@MaximilianAlgehed I have updated the comment, and ensured `suchThat` is used in the generator produced by `withGenQ`. This lead me to add a property test which I think could be generalised as a way to ensure `Quantification`s are consistent.

**Comment by MaximilianAlgehed** [comment]
@abailly-iohk I like it! I agree, we should add more of those tests for the builtin quantifications as regression tests to make sure we don't break anything in the future.

**Review by MaximilianAlgehed — APPROVED** [review]

**Comment by abailly-iohk** [comment]
And it's also a way to document the behaviour of the system in a more precise and unambiguous way than with words.

**Comment by MaximilianAlgehed** [comment]
About time to merge this right?

**Pull request merged**